From: petronijevicm <71526010+petronijevicm@users.noreply.github.com>
Subject: [PATCH] Settings: Motion Assist accessibility screens

Adds Accessibility > Motion cues: animated tutorial, main switch,
start-in-vehicle, Quick Settings tile shortcut, and a Customize page
(colour, shape, opacity, size, movement, responsiveness, display area,
shuffle, smooth animation). Values are stored in Settings.Secure
motion_assist_* keys read by SystemUI (see the frameworks/base patch).

SPDX-License-Identifier: Apache-2.0
---
 res/drawable/ic_motion_assist.xml                  |  11 +
 res/drawable/ic_motion_assist_edit.xml             |  11 +
 res/drawable/motion_assist_tutorial.xml            | 137 +++++++++++++
 res/values-es/motion_assist_strings.xml            |  51 +++++
 res/values/motion_assist_arrays.xml                |  13 ++
 res/values/motion_assist_strings.xml               |  51 +++++
 res/xml/accessibility_settings.xml                 |   9 +
 res/xml/motion_assist_customize_settings.xml       |  70 +++++++
 res/xml/motion_assist_settings.xml                 |  54 +++++
 .../MotionAssistChoicePreferenceController.java    | 228 +++++++++++++++++++++
 .../MotionAssistCustomizeSettings.java             |  38 ++++
 ...otionAssistDisplayAreaPreferenceController.java |  49 +++++
 .../MotionAssistEntryPreferenceController.java     |  34 +++
 .../motionassist/MotionAssistKeys.java             |  47 +++++
 .../motionassist/MotionAssistSettings.java         |  38 ++++
 .../MotionAssistSliderPreferenceController.java    |  85 ++++++++
 .../MotionAssistTilePreferenceController.java      |  83 ++++++++
 .../MotionAssistTogglePreferenceController.java    |  63 ++++++
 18 files changed, 1072 insertions(+)

diff --git a/res/drawable/ic_motion_assist.xml b/res/drawable/ic_motion_assist.xml
new file mode 100644
index 0000000..b343338
--- /dev/null
+++ b/res/drawable/ic_motion_assist.xml
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!-- SPDX-FileCopyrightText: 2026 petronijevicm -->
+<!-- SPDX-License-Identifier: Apache-2.0 -->
+<vector xmlns:android="http://schemas.android.com/apk/res/android"
+    android:width="24dp"
+    android:height="24dp"
+    android:viewportWidth="24"
+    android:viewportHeight="24"
+    android:tint="?android:attr/colorControlNormal">
+    <path android:fillColor="#FFFFFFFF" android:fillType="evenOdd" android:pathData="M9.7,2.5 H14.3 A2.2,2.2 0 0 1 16.5,4.7 V19.3 A2.2,2.2 0 0 1 14.3,21.5 H9.7 A2.2,2.2 0 0 1 7.5,19.3 V4.7 A2.2,2.2 0 0 1 9.7,2.5 Z M10.200000000000001,4.3 H13.8 A0.9,0.9 0 0 1 14.700000000000001,5.2 V18.8 A0.9,0.9 0 0 1 13.8,19.7 H10.200000000000001 A0.9,0.9 0 0 1 9.3,18.8 V5.2 A0.9,0.9 0 0 1 10.200000000000001,4.3 Z M1.7000000000000002,7 a1.5,1.5 0 1,0 3.0,0 a1.5,1.5 0 1,0 -3.0,0 Z M1.7000000000000002,12 a1.5,1.5 0 1,0 3.0,0 a1.5,1.5 0 1,0 -3.0,0 Z M1.7000000000000002,17 a1.5,1.5 0 1,0 3.0,0 a1.5,1.5 0 1,0 -3.0,0 Z M19.3,7 a1.5,1.5 0 1,0 3.0,0 a1.5,1.5 0 1,0 -3.0,0 Z M19.3,12 a1.5,1.5 0 1,0 3.0,0 a1.5,1.5 0 1,0 -3.0,0 Z M19.3,17 a1.5,1.5 0 1,0 3.0,0 a1.5,1.5 0 1,0 -3.0,0 Z" />
+</vector>
diff --git a/res/drawable/ic_motion_assist_edit.xml b/res/drawable/ic_motion_assist_edit.xml
new file mode 100644
index 0000000..515e8cb
--- /dev/null
+++ b/res/drawable/ic_motion_assist_edit.xml
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!-- SPDX-FileCopyrightText: 2026 petronijevicm -->
+<!-- SPDX-License-Identifier: Apache-2.0 -->
+<vector xmlns:android="http://schemas.android.com/apk/res/android"
+    android:width="24dp"
+    android:height="24dp"
+    android:viewportWidth="24"
+    android:viewportHeight="24"
+    android:tint="?android:attr/colorControlNormal">
+    <path android:fillColor="#FFFFFFFF" android:pathData="M4,17.5 L4,20 L6.5,20 L16.8,9.7 L14.3,7.2 Z M18.2,8.3 L19.6,6.9 C20,6.5 20,5.9 19.6,5.5 L18.5,4.4 C18.1,4 17.5,4 17.1,4.4 L15.7,5.8 Z" />
+</vector>
diff --git a/res/drawable/motion_assist_tutorial.xml b/res/drawable/motion_assist_tutorial.xml
new file mode 100644
index 0000000..77e2c06
--- /dev/null
+++ b/res/drawable/motion_assist_tutorial.xml
@@ -0,0 +1,137 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!--
+  SPDX-FileCopyrightText: 2026 petronijevicm
+  SPDX-License-Identifier: Apache-2.0
+
+  Generated by tools/make_tutorial_drawable.py - edit the script, not this file.
+-->
+<animated-vector
+    xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:aapt="http://schemas.android.com/aapt">
+
+    <aapt:attr name="android:drawable">
+        <vector
+            android:width="360dp"
+            android:height="200dp"
+            android:viewportWidth="360"
+            android:viewportHeight="200">
+
+            <!-- Steering wheel -->
+            <group android:name="wheel" android:pivotX="66" android:pivotY="100">
+                <path android:strokeColor="?android:attr/textColorSecondary" android:strokeWidth="6" android:pathData="M36,100 a30,30 0 1,0 60,0 a30,30 0 1,0 -60,0 Z" />
+                <path android:fillColor="?android:attr/textColorSecondary" android:pathData="M59,100 a7,7 0 1,0 14,0 a7,7 0 1,0 -14,0 Z" />
+            <path android:fillColor="?android:attr/textColorSecondary" android:pathData="M36,97.5 H60 V102.5 H36 Z" />
+            <path android:fillColor="?android:attr/textColorSecondary" android:pathData="M72,97.5 H96 V102.5 H72 Z" />
+            <path android:fillColor="?android:attr/textColorSecondary" android:pathData="M63.5,106 H68.5 V130 H63.5 Z" />
+                <path android:fillColor="?android:attr/colorAccent" android:pathData="M64,66 H68 A2,2 0 0 1 70,68 V72 A2,2 0 0 1 68,74 H64 A2,2 0 0 1 62,72 V68 A2,2 0 0 1 64,66 Z" />
+            </group>
+
+            <!-- Phone -->
+            <path android:strokeColor="?android:attr/textColorSecondary" android:strokeWidth="3" android:pathData="M146,12 H214 A14,14 0 0 1 228,26 V174 A14,14 0 0 1 214,188 H146 A14,14 0 0 1 132,174 V26 A14,14 0 0 1 146,12 Z" />
+            <group android:name="screen">
+                <clip-path android:pathData="M146,22 H214 A8,8 0 0 1 222,30 V170 A8,8 0 0 1 214,178 H146 A8,8 0 0 1 138,170 V30 A8,8 0 0 1 146,22 Z" />
+                <path android:fillColor="?android:attr/colorBackgroundFloating" android:pathData="M146,22 H214 A8,8 0 0 1 222,30 V170 A8,8 0 0 1 214,178 H146 A8,8 0 0 1 138,170 V30 A8,8 0 0 1 146,22 Z" />
+                <path android:fillColor="?android:attr/textColorTertiary" android:pathData="M169,58 H191 A3,3 0 0 1 194,61 V61 A3,3 0 0 1 191,64 H169 A3,3 0 0 1 166,61 V61 A3,3 0 0 1 169,58 Z" />
+                <path android:fillColor="?android:attr/textColorTertiary" android:pathData="M169,72 H185 A3,3 0 0 1 188,75 V75 A3,3 0 0 1 185,78 H169 A3,3 0 0 1 166,75 V75 A3,3 0 0 1 169,72 Z" />
+                <path android:fillColor="?android:attr/textColorTertiary" android:pathData="M169,86 H191 A3,3 0 0 1 194,89 V89 A3,3 0 0 1 191,92 H169 A3,3 0 0 1 166,89 V89 A3,3 0 0 1 169,86 Z" />
+                <path android:fillColor="?android:attr/textColorTertiary" android:pathData="M169,100 H181 A3,3 0 0 1 184,103 V103 A3,3 0 0 1 181,106 H169 A3,3 0 0 1 166,103 V103 A3,3 0 0 1 169,100 Z" />
+                <path android:fillColor="?android:attr/textColorTertiary" android:pathData="M169,114 H191 A3,3 0 0 1 194,117 V117 A3,3 0 0 1 191,120 H169 A3,3 0 0 1 166,117 V117 A3,3 0 0 1 169,114 Z" />
+                <path android:fillColor="?android:attr/textColorTertiary" android:pathData="M169,128 H187 A3,3 0 0 1 190,131 V131 A3,3 0 0 1 187,134 H169 A3,3 0 0 1 166,131 V131 A3,3 0 0 1 169,128 Z" />
+                <path android:fillColor="?android:attr/textColorTertiary" android:pathData="M169,142 H177 A3,3 0 0 1 180,145 V145 A3,3 0 0 1 177,148 H169 A3,3 0 0 1 166,145 V145 A3,3 0 0 1 169,142 Z" />
+                <group android:name="cues">
+                    <path android:fillColor="?android:attr/colorAccent" android:pathData="M141,-2 a5,5 0 1,0 10,0 a5,5 0 1,0 -10,0 Z" />
+                    <path android:fillColor="?android:attr/colorAccent" android:pathData="M149,20 a5,5 0 1,0 10,0 a5,5 0 1,0 -10,0 Z" />
+                    <path android:fillColor="?android:attr/colorAccent" android:pathData="M141,42 a5,5 0 1,0 10,0 a5,5 0 1,0 -10,0 Z" />
+                    <path android:fillColor="?android:attr/colorAccent" android:pathData="M149,64 a5,5 0 1,0 10,0 a5,5 0 1,0 -10,0 Z" />
+                    <path android:fillColor="?android:attr/colorAccent" android:pathData="M141,86 a5,5 0 1,0 10,0 a5,5 0 1,0 -10,0 Z" />
+                    <path android:fillColor="?android:attr/colorAccent" android:pathData="M149,108 a5,5 0 1,0 10,0 a5,5 0 1,0 -10,0 Z" />
+                    <path android:fillColor="?android:attr/colorAccent" android:pathData="M141,130 a5,5 0 1,0 10,0 a5,5 0 1,0 -10,0 Z" />
+                    <path android:fillColor="?android:attr/colorAccent" android:pathData="M149,152 a5,5 0 1,0 10,0 a5,5 0 1,0 -10,0 Z" />
+                    <path android:fillColor="?android:attr/colorAccent" android:pathData="M141,174 a5,5 0 1,0 10,0 a5,5 0 1,0 -10,0 Z" />
+                    <path android:fillColor="?android:attr/colorAccent" android:pathData="M149,196 a5,5 0 1,0 10,0 a5,5 0 1,0 -10,0 Z" />
+                    <path android:fillColor="?android:attr/colorAccent" android:pathData="M209,-2 a5,5 0 1,0 10,0 a5,5 0 1,0 -10,0 Z" />
+                    <path android:fillColor="?android:attr/colorAccent" android:pathData="M201,20 a5,5 0 1,0 10,0 a5,5 0 1,0 -10,0 Z" />
+                    <path android:fillColor="?android:attr/colorAccent" android:pathData="M209,42 a5,5 0 1,0 10,0 a5,5 0 1,0 -10,0 Z" />
+                    <path android:fillColor="?android:attr/colorAccent" android:pathData="M201,64 a5,5 0 1,0 10,0 a5,5 0 1,0 -10,0 Z" />
+                    <path android:fillColor="?android:attr/colorAccent" android:pathData="M209,86 a5,5 0 1,0 10,0 a5,5 0 1,0 -10,0 Z" />
+                    <path android:fillColor="?android:attr/colorAccent" android:pathData="M201,108 a5,5 0 1,0 10,0 a5,5 0 1,0 -10,0 Z" />
+                    <path android:fillColor="?android:attr/colorAccent" android:pathData="M209,130 a5,5 0 1,0 10,0 a5,5 0 1,0 -10,0 Z" />
+                    <path android:fillColor="?android:attr/colorAccent" android:pathData="M201,152 a5,5 0 1,0 10,0 a5,5 0 1,0 -10,0 Z" />
+                    <path android:fillColor="?android:attr/colorAccent" android:pathData="M209,174 a5,5 0 1,0 10,0 a5,5 0 1,0 -10,0 Z" />
+                    <path android:fillColor="?android:attr/colorAccent" android:pathData="M201,196 a5,5 0 1,0 10,0 a5,5 0 1,0 -10,0 Z" />
+                </group>
+            </group>
+
+            <!-- Brake pedal -->
+            <group android:name="pedal" android:pivotX="294" android:pivotY="124">
+                <path android:fillColor="?android:attr/colorAccent" android:pathData="M286,76 H302 A6,6 0 0 1 308,82 V118 A6,6 0 0 1 302,124 H286 A6,6 0 0 1 280,118 V82 A6,6 0 0 1 286,76 Z" />
+                <path android:fillColor="?android:attr/colorBackgroundFloating" android:pathData="M288,86 H300 A2,2 0 0 1 302,88 V88 A2,2 0 0 1 300,90 H288 A2,2 0 0 1 286,88 V88 A2,2 0 0 1 288,86 Z" />
+                <path android:fillColor="?android:attr/colorBackgroundFloating" android:pathData="M288,96 H300 A2,2 0 0 1 302,98 V98 A2,2 0 0 1 300,100 H288 A2,2 0 0 1 286,98 V98 A2,2 0 0 1 288,96 Z" />
+                <path android:fillColor="?android:attr/colorBackgroundFloating" android:pathData="M288,106 H300 A2,2 0 0 1 302,108 V108 A2,2 0 0 1 300,110 H288 A2,2 0 0 1 286,108 V108 A2,2 0 0 1 288,106 Z" />
+            </group>
+        </vector>
+    </aapt:attr>
+
+    <target android:name="wheel">
+        <aapt:attr name="android:animation">
+            <objectAnimator
+                android:duration="6000"
+                android:repeatCount="infinite">
+                    <propertyValuesHolder android:propertyName="rotation">
+                        <keyframe android:fraction="0" android:value="0" android:valueType="floatType" />
+                        <keyframe android:fraction="0.07" android:value="0" android:valueType="floatType" android:interpolator="@android:interpolator/fast_out_slow_in" />
+                        <keyframe android:fraction="0.23" android:value="-35" android:valueType="floatType" android:interpolator="@android:interpolator/fast_out_slow_in" />
+                        <keyframe android:fraction="0.3" android:value="-35" android:valueType="floatType" android:interpolator="@android:interpolator/fast_out_slow_in" />
+                        <keyframe android:fraction="0.4" android:value="0" android:valueType="floatType" android:interpolator="@android:interpolator/fast_out_slow_in" />
+                        <keyframe android:fraction="0.53" android:value="35" android:valueType="floatType" android:interpolator="@android:interpolator/fast_out_slow_in" />
+                        <keyframe android:fraction="0.6" android:value="35" android:valueType="floatType" android:interpolator="@android:interpolator/fast_out_slow_in" />
+                        <keyframe android:fraction="0.7" android:value="0" android:valueType="floatType" android:interpolator="@android:interpolator/fast_out_slow_in" />
+                        <keyframe android:fraction="1" android:value="0" android:valueType="floatType" android:interpolator="@android:interpolator/fast_out_slow_in" />
+                    </propertyValuesHolder>
+            </objectAnimator>
+        </aapt:attr>
+    </target>
+    <target android:name="cues">
+        <aapt:attr name="android:animation">
+            <objectAnimator
+                android:duration="6000"
+                android:repeatCount="infinite">
+                    <propertyValuesHolder android:propertyName="translateX">
+                        <keyframe android:fraction="0" android:value="0" android:valueType="floatType" />
+                        <keyframe android:fraction="0.07" android:value="0" android:valueType="floatType" android:interpolator="@android:interpolator/fast_out_slow_in" />
+                        <keyframe android:fraction="0.23" android:value="8" android:valueType="floatType" android:interpolator="@android:interpolator/fast_out_slow_in" />
+                        <keyframe android:fraction="0.3" android:value="8" android:valueType="floatType" android:interpolator="@android:interpolator/fast_out_slow_in" />
+                        <keyframe android:fraction="0.4" android:value="0" android:valueType="floatType" android:interpolator="@android:interpolator/fast_out_slow_in" />
+                        <keyframe android:fraction="0.53" android:value="-8" android:valueType="floatType" android:interpolator="@android:interpolator/fast_out_slow_in" />
+                        <keyframe android:fraction="0.6" android:value="-8" android:valueType="floatType" android:interpolator="@android:interpolator/fast_out_slow_in" />
+                        <keyframe android:fraction="0.7" android:value="0" android:valueType="floatType" android:interpolator="@android:interpolator/fast_out_slow_in" />
+                        <keyframe android:fraction="1" android:value="0" android:valueType="floatType" android:interpolator="@android:interpolator/fast_out_slow_in" />
+                    </propertyValuesHolder>
+                    <propertyValuesHolder android:propertyName="translateY">
+                        <keyframe android:fraction="0" android:value="0" android:valueType="floatType" />
+                        <keyframe android:fraction="0.72" android:value="0" android:valueType="floatType" android:interpolator="@android:interpolator/fast_out_slow_in" />
+                        <keyframe android:fraction="0.78" android:value="-12" android:valueType="floatType" android:interpolator="@android:interpolator/fast_out_slow_in" />
+                        <keyframe android:fraction="0.85" android:value="-12" android:valueType="floatType" android:interpolator="@android:interpolator/fast_out_slow_in" />
+                        <keyframe android:fraction="0.95" android:value="0" android:valueType="floatType" android:interpolator="@android:interpolator/fast_out_slow_in" />
+                        <keyframe android:fraction="1" android:value="0" android:valueType="floatType" android:interpolator="@android:interpolator/fast_out_slow_in" />
+                    </propertyValuesHolder>
+            </objectAnimator>
+        </aapt:attr>
+    </target>
+    <target android:name="pedal">
+        <aapt:attr name="android:animation">
+            <objectAnimator
+                android:duration="6000"
+                android:repeatCount="infinite">
+                    <propertyValuesHolder android:propertyName="scaleY">
+                        <keyframe android:fraction="0" android:value="1" android:valueType="floatType" />
+                        <keyframe android:fraction="0.72" android:value="1" android:valueType="floatType" android:interpolator="@android:interpolator/fast_out_slow_in" />
+                        <keyframe android:fraction="0.78" android:value="0.78" android:valueType="floatType" android:interpolator="@android:interpolator/fast_out_slow_in" />
+                        <keyframe android:fraction="0.85" android:value="0.78" android:valueType="floatType" android:interpolator="@android:interpolator/fast_out_slow_in" />
+                        <keyframe android:fraction="0.95" android:value="1" android:valueType="floatType" android:interpolator="@android:interpolator/fast_out_slow_in" />
+                        <keyframe android:fraction="1" android:value="1" android:valueType="floatType" android:interpolator="@android:interpolator/fast_out_slow_in" />
+                    </propertyValuesHolder>
+            </objectAnimator>
+        </aapt:attr>
+    </target>
+</animated-vector>
diff --git a/res/values-es/motion_assist_strings.xml b/res/values-es/motion_assist_strings.xml
new file mode 100644
index 0000000..1d31761
--- /dev/null
+++ b/res/values-es/motion_assist_strings.xml
@@ -0,0 +1,51 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!-- SPDX-FileCopyrightText: 2026 petronijevicm -->
+<!-- SPDX-License-Identifier: Apache-2.0 -->
+<resources>
+    <string name="motion_assist_title">Señales de movimiento</string>
+    <string name="motion_assist_intro">Los puntos que se mueven en los bordes de la pantalla siguen el movimiento del vehículo. Esto puede aliviar el mareo mientras lees o ves vídeos en un coche, autobús o tren.</string>
+    <string name="motion_assist_main_switch">Usar señales de movimiento</string>
+    <string name="motion_assist_auto_start_title">Iniciar en un vehículo en movimiento</string>
+    <string name="motion_assist_auto_start_summary">Mostrar las señales solo cuando se detecte el movimiento del vehículo</string>
+    <string name="motion_assist_options">Opciones</string>
+    <string name="motion_assist_tile_title">Añadir botón a Ajustes rápidos</string>
+    <string name="motion_assist_tile_summary">Activa o desactiva las señales desde cualquier pantalla</string>
+    <string name="motion_assist_customize_title">Personalizar</string>
+    <string name="motion_assist_customize_summary">Color, forma, tamaño y movimiento</string>
+    <string name="motion_assist_disclaimer">Las señales de movimiento pueden reducir el mareo, pero no lo evitan en todas las personas.</string>
+    <string name="motion_assist_customize_intro">Los cambios se ven al instante mientras las señales están activas.</string>
+    <string name="motion_assist_color_title">Color</string>
+    <string name="motion_assist_color_system">Fondo de pantalla</string>
+    <string name="motion_assist_color_salmon">Salmón</string>
+    <string name="motion_assist_color_amber">Ámbar</string>
+    <string name="motion_assist_color_mint">Menta</string>
+    <string name="motion_assist_color_sky">Cielo</string>
+    <string name="motion_assist_color_contrast">Alto contraste</string>
+    <string name="motion_assist_shape_title">Forma</string>
+    <string name="motion_assist_shape_circle">Círculo</string>
+    <string name="motion_assist_shape_squircle">Squircle</string>
+    <string name="motion_assist_shape_pentagon">Pentágono</string>
+    <string name="motion_assist_shape_diamond">Rombo</string>
+    <string name="motion_assist_opacity_title">Opacidad</string>
+    <string name="motion_assist_size_title">Tamaño</string>
+    <string name="motion_assist_movement_title">Movimiento</string>
+    <string name="motion_assist_movement_summary">Cuánto se desplazan las señales cuando el vehículo acelera, frena o gira</string>
+    <string name="motion_assist_responsiveness_title">Capacidad de respuesta</string>
+    <string name="motion_assist_responsiveness_summary">Más bajo es más suave y tranquilo, más alto reacciona antes</string>
+    <string name="motion_assist_area_title">Zona de visualización</string>
+    <string name="motion_assist_area_sides">Franjas laterales (recomendado)</string>
+    <string name="motion_assist_area_full">Pantalla completa</string>
+    <string name="motion_assist_shuffle_title">Mezclar</string>
+    <string name="motion_assist_shuffle_summary">Dispersa los puntos y cambia la forma y el color cada pocos segundos</string>
+    <string name="motion_assist_smooth_title">Animación fluida</string>
+    <string name="motion_assist_smooth_summary">Dibuja a la frecuencia de actualización máxima de la pantalla. Limitada con Ahorro de batería</string>
+    <string name="motion_assist_save">Guardar</string>
+    <string name="motion_assist_got_it">Entendido</string>
+    <string name="motion_assist_tile_dialog_title">Botón añadido</string>
+    <string name="motion_assist_tile_dialog_use">Desliza hacia abajo para abrir Ajustes rápidos y toca Señales de movimiento para activarlas o desactivarlas. Mantén pulsado el botón para abrir estos ajustes.</string>
+    <string name="motion_assist_tile_dialog_edit" formatted="false">Para mover el botón, toca %s y arrástralo adonde quieras.</string>
+    <string name="motion_assist_state_on">Activadas</string>
+    <string name="motion_assist_state_off">Desactivadas</string>
+    <string name="motion_assist_state_auto">Activadas en vehículos en movimiento</string>
+    <string name="motion_assist_keywords">mareo, mareo en coche, vehículo, señales, náuseas</string>
+</resources>
diff --git a/res/values/motion_assist_arrays.xml b/res/values/motion_assist_arrays.xml
new file mode 100644
index 0000000..27c7f8a
--- /dev/null
+++ b/res/values/motion_assist_arrays.xml
@@ -0,0 +1,13 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!-- SPDX-FileCopyrightText: 2026 petronijevicm -->
+<!-- SPDX-License-Identifier: Apache-2.0 -->
+<resources>
+    <string-array name="motion_assist_area_entries">
+        <item>@string/motion_assist_area_sides</item>
+        <item>@string/motion_assist_area_full</item>
+    </string-array>
+    <string-array name="motion_assist_area_values" translatable="false">
+        <item>0</item>
+        <item>1</item>
+    </string-array>
+</resources>
diff --git a/res/values/motion_assist_strings.xml b/res/values/motion_assist_strings.xml
new file mode 100644
index 0000000..46530ff
--- /dev/null
+++ b/res/values/motion_assist_strings.xml
@@ -0,0 +1,51 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!-- SPDX-FileCopyrightText: 2026 petronijevicm -->
+<!-- SPDX-License-Identifier: Apache-2.0 -->
+<resources>
+    <string name="motion_assist_title">Motion cues</string>
+    <string name="motion_assist_intro">Moving dots at the edges of the screen follow the vehicle\'s motion. This can ease motion sickness while you read or watch in a car, bus or train.</string>
+    <string name="motion_assist_main_switch">Use motion cues</string>
+    <string name="motion_assist_auto_start_title">Start in a moving vehicle</string>
+    <string name="motion_assist_auto_start_summary">Show cues only while vehicle motion is detected</string>
+    <string name="motion_assist_options">Options</string>
+    <string name="motion_assist_tile_title">Add tile to Quick Settings</string>
+    <string name="motion_assist_tile_summary">Turn cues on or off from any screen</string>
+    <string name="motion_assist_customize_title">Customize</string>
+    <string name="motion_assist_customize_summary">Color, shape, size and movement</string>
+    <string name="motion_assist_disclaimer">Motion cues may reduce motion sickness, but they don\'t prevent it for everyone.</string>
+    <string name="motion_assist_customize_intro">Changes show right away while cues are on.</string>
+    <string name="motion_assist_color_title">Color</string>
+    <string name="motion_assist_color_system">Wallpaper</string>
+    <string name="motion_assist_color_salmon">Salmon</string>
+    <string name="motion_assist_color_amber">Amber</string>
+    <string name="motion_assist_color_mint">Mint</string>
+    <string name="motion_assist_color_sky">Sky</string>
+    <string name="motion_assist_color_contrast">High contrast</string>
+    <string name="motion_assist_shape_title">Shape</string>
+    <string name="motion_assist_shape_circle">Circle</string>
+    <string name="motion_assist_shape_squircle">Squircle</string>
+    <string name="motion_assist_shape_pentagon">Pentagon</string>
+    <string name="motion_assist_shape_diamond">Diamond</string>
+    <string name="motion_assist_opacity_title">Opacity</string>
+    <string name="motion_assist_size_title">Size</string>
+    <string name="motion_assist_movement_title">Movement</string>
+    <string name="motion_assist_movement_summary">How far the cues travel when the vehicle speeds up, brakes or turns</string>
+    <string name="motion_assist_responsiveness_title">Responsiveness</string>
+    <string name="motion_assist_responsiveness_summary">Lower is calmer and smoother, higher reacts sooner</string>
+    <string name="motion_assist_area_title">Display area</string>
+    <string name="motion_assist_area_sides">Side bands (recommended)</string>
+    <string name="motion_assist_area_full">Full screen</string>
+    <string name="motion_assist_shuffle_title">Shuffle</string>
+    <string name="motion_assist_shuffle_summary">Scatter the dots and change shape and color every few seconds</string>
+    <string name="motion_assist_smooth_title">Smooth animation</string>
+    <string name="motion_assist_smooth_summary">Draw at the full refresh rate of your display. Limited in Battery Saver</string>
+    <string name="motion_assist_save">Save</string>
+    <string name="motion_assist_got_it">Got it</string>
+    <string name="motion_assist_tile_dialog_title">Tile added</string>
+    <string name="motion_assist_tile_dialog_use">Swipe down to open Quick Settings and tap Motion cues to turn them on or off. Long-press the tile to open these settings.</string>
+    <string name="motion_assist_tile_dialog_edit" formatted="false">To move the tile, tap %s and drag it where you want.</string>
+    <string name="motion_assist_state_on">On</string>
+    <string name="motion_assist_state_off">Off</string>
+    <string name="motion_assist_state_auto">On in moving vehicles</string>
+    <string name="motion_assist_keywords">motion sickness, car sickness, travel sickness, vehicle, cues, nausea</string>
+</resources>
diff --git a/res/xml/accessibility_settings.xml b/res/xml/accessibility_settings.xml
index b0ccb3c..cff9088 100644
--- a/res/xml/accessibility_settings.xml
+++ b/res/xml/accessibility_settings.xml
@@ -54,6 +54,15 @@
             android:title="@string/accessibility_color_and_motion_title"
             settings:searchable="true"/>
 
+        <Preference
+            android:fragment="com.android.settings.accessibility.motionassist.MotionAssistSettings"
+            android:key="motion_assist_preference_screen"
+            android:icon="@drawable/ic_motion_assist"
+            android:persistent="false"
+            android:title="@string/motion_assist_title"
+            settings:keywords="@string/motion_assist_keywords"
+            settings:controller="com.android.settings.accessibility.motionassist.MotionAssistEntryPreferenceController"/>
+
         <com.android.settingslib.PrimarySwitchPreference
             android:fragment="com.android.settings.accessibility.ToggleReduceBrightColorsPreferenceFragment"
             android:key="reduce_bright_colors_preference"
diff --git a/res/xml/motion_assist_customize_settings.xml b/res/xml/motion_assist_customize_settings.xml
new file mode 100644
index 0000000..6cc5aed
--- /dev/null
+++ b/res/xml/motion_assist_customize_settings.xml
@@ -0,0 +1,70 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!-- SPDX-FileCopyrightText: 2026 petronijevicm -->
+<!-- SPDX-License-Identifier: Apache-2.0 -->
+<PreferenceScreen
+    xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:settings="http://schemas.android.com/apk/res-auto"
+    android:key="motion_assist_customize_screen"
+    android:title="@string/motion_assist_customize_title">
+
+    <com.android.settingslib.widget.TopIntroPreference
+        android:key="motion_assist_customize_intro"
+        android:title="@string/motion_assist_customize_intro"
+        settings:searchable="false" />
+
+    <Preference
+        android:key="motion_assist_color"
+        android:title="@string/motion_assist_color_title"
+        settings:controller="com.android.settings.accessibility.motionassist.MotionAssistChoicePreferenceController" />
+
+    <Preference
+        android:key="motion_assist_shape"
+        android:title="@string/motion_assist_shape_title"
+        settings:controller="com.android.settings.accessibility.motionassist.MotionAssistChoicePreferenceController" />
+
+    <com.android.settingslib.widget.SliderPreference
+        android:key="motion_assist_opacity"
+        android:persistent="false"
+        android:title="@string/motion_assist_opacity_title"
+        settings:controller="com.android.settings.accessibility.motionassist.MotionAssistSliderPreferenceController" />
+
+    <com.android.settingslib.widget.SliderPreference
+        android:key="motion_assist_size"
+        android:persistent="false"
+        android:title="@string/motion_assist_size_title"
+        settings:controller="com.android.settings.accessibility.motionassist.MotionAssistSliderPreferenceController" />
+
+    <com.android.settingslib.widget.SliderPreference
+        android:key="motion_assist_movement"
+        android:persistent="false"
+        android:title="@string/motion_assist_movement_title"
+        android:summary="@string/motion_assist_movement_summary"
+        settings:controller="com.android.settings.accessibility.motionassist.MotionAssistSliderPreferenceController" />
+
+    <com.android.settingslib.widget.SliderPreference
+        android:key="motion_assist_responsiveness"
+        android:persistent="false"
+        android:title="@string/motion_assist_responsiveness_title"
+        android:summary="@string/motion_assist_responsiveness_summary"
+        settings:controller="com.android.settings.accessibility.motionassist.MotionAssistSliderPreferenceController" />
+
+    <ListPreference
+        android:key="motion_assist_display_area"
+        android:persistent="false"
+        android:title="@string/motion_assist_area_title"
+        android:entries="@array/motion_assist_area_entries"
+        android:entryValues="@array/motion_assist_area_values"
+        settings:controller="com.android.settings.accessibility.motionassist.MotionAssistDisplayAreaPreferenceController" />
+
+    <SwitchPreferenceCompat
+        android:key="motion_assist_shuffle"
+        android:title="@string/motion_assist_shuffle_title"
+        android:summary="@string/motion_assist_shuffle_summary"
+        settings:controller="com.android.settings.accessibility.motionassist.MotionAssistTogglePreferenceController" />
+
+    <SwitchPreferenceCompat
+        android:key="motion_assist_smooth"
+        android:title="@string/motion_assist_smooth_title"
+        android:summary="@string/motion_assist_smooth_summary"
+        settings:controller="com.android.settings.accessibility.motionassist.MotionAssistTogglePreferenceController" />
+</PreferenceScreen>
diff --git a/res/xml/motion_assist_settings.xml b/res/xml/motion_assist_settings.xml
new file mode 100644
index 0000000..649e63e
--- /dev/null
+++ b/res/xml/motion_assist_settings.xml
@@ -0,0 +1,54 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!-- SPDX-FileCopyrightText: 2026 petronijevicm -->
+<!-- SPDX-License-Identifier: Apache-2.0 -->
+<PreferenceScreen
+    xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:settings="http://schemas.android.com/apk/res-auto"
+    android:key="motion_assist_screen"
+    android:title="@string/motion_assist_title"
+    settings:keywords="@string/motion_assist_keywords">
+
+    <com.android.settingslib.widget.TopIntroPreference
+        android:key="motion_assist_intro"
+        android:title="@string/motion_assist_intro"
+        settings:searchable="false" />
+
+    <com.android.settingslib.widget.IllustrationPreference
+        android:key="motion_assist_illustration"
+        android:selectable="false"
+        settings:searchable="false"
+        settings:lottie_rawRes="@drawable/motion_assist_tutorial" />
+
+    <com.android.settingslib.widget.MainSwitchPreference
+        android:key="motion_assist_main_switch"
+        android:title="@string/motion_assist_main_switch"
+        settings:controller="com.android.settings.accessibility.motionassist.MotionAssistTogglePreferenceController" />
+
+    <SwitchPreferenceCompat
+        android:key="motion_assist_auto_start"
+        android:title="@string/motion_assist_auto_start_title"
+        android:summary="@string/motion_assist_auto_start_summary"
+        settings:controller="com.android.settings.accessibility.motionassist.MotionAssistTogglePreferenceController" />
+
+    <PreferenceCategory
+        android:key="motion_assist_options"
+        android:title="@string/motion_assist_options">
+
+        <Preference
+            android:key="motion_assist_add_tile"
+            android:title="@string/motion_assist_tile_title"
+            android:summary="@string/motion_assist_tile_summary"
+            settings:controller="com.android.settings.accessibility.motionassist.MotionAssistTilePreferenceController" />
+
+        <Preference
+            android:key="motion_assist_customize"
+            android:title="@string/motion_assist_customize_title"
+            android:summary="@string/motion_assist_customize_summary"
+            android:fragment="com.android.settings.accessibility.motionassist.MotionAssistCustomizeSettings" />
+    </PreferenceCategory>
+
+    <com.android.settingslib.widget.FooterPreference
+        android:key="motion_assist_footer"
+        android:title="@string/motion_assist_disclaimer"
+        settings:searchable="false" />
+</PreferenceScreen>
diff --git a/src/com/android/settings/accessibility/motionassist/MotionAssistChoicePreferenceController.java b/src/com/android/settings/accessibility/motionassist/MotionAssistChoicePreferenceController.java
new file mode 100644
index 0000000..c37b604
--- /dev/null
+++ b/src/com/android/settings/accessibility/motionassist/MotionAssistChoicePreferenceController.java
@@ -0,0 +1,228 @@
+/*
+ * SPDX-FileCopyrightText: 2026 petronijevicm
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package com.android.settings.accessibility.motionassist;
+
+import android.content.Context;
+import android.content.res.Configuration;
+import android.graphics.Canvas;
+import android.graphics.Color;
+import android.graphics.Paint;
+import android.graphics.Path;
+import android.graphics.drawable.GradientDrawable;
+import android.text.TextUtils;
+import android.util.TypedValue;
+import android.view.Gravity;
+import android.view.View;
+import android.widget.FrameLayout;
+import android.widget.LinearLayout;
+
+import androidx.appcompat.app.AlertDialog;
+import androidx.preference.Preference;
+
+import com.android.settings.R;
+import com.android.settings.core.BasePreferenceController;
+
+/**
+ * Colour and shape pickers. Each shows a row of swatches; picking one applies it at once so
+ * running cues update live, Save keeps it and Cancel restores the previous choice.
+ */
+public class MotionAssistChoicePreferenceController extends BasePreferenceController {
+
+    private static final String KEY_COLOR = "motion_assist_color";
+    private static final String KEY_SHAPE = "motion_assist_shape";
+
+    /** Index 0 (wallpaper accent) and the last (high contrast) are resolved at runtime. */
+    private static final int[] PALETTE = {
+            0,
+            Color.rgb(0xF2, 0x9C, 0x86),
+            Color.rgb(0xF5, 0xB7, 0x2A),
+            Color.rgb(0x7C, 0xC9, 0x96),
+            Color.rgb(0x86, 0xBC, 0xEE),
+            0,
+    };
+    private static final int[] COLOR_NAMES = {
+            R.string.motion_assist_color_system,
+            R.string.motion_assist_color_salmon,
+            R.string.motion_assist_color_amber,
+            R.string.motion_assist_color_mint,
+            R.string.motion_assist_color_sky,
+            R.string.motion_assist_color_contrast,
+    };
+    private static final int[] SHAPE_NAMES = {
+            R.string.motion_assist_shape_circle,
+            R.string.motion_assist_shape_squircle,
+            R.string.motion_assist_shape_pentagon,
+            R.string.motion_assist_shape_diamond,
+    };
+
+    public MotionAssistChoicePreferenceController(Context context, String key) {
+        super(context, key);
+    }
+
+    @Override
+    public int getAvailabilityStatus() {
+        return isColor() || isShape() ? AVAILABLE : UNSUPPORTED_ON_DEVICE;
+    }
+
+    @Override
+    public CharSequence getSummary() {
+        final int[] names = isColor() ? COLOR_NAMES : SHAPE_NAMES;
+        return mContext.getString(names[clamp(current(), names.length)]);
+    }
+
+    @Override
+    public boolean handlePreferenceTreeClick(Preference preference) {
+        if (!TextUtils.equals(preference.getKey(), getPreferenceKey())) {
+            return super.handlePreferenceTreeClick(preference);
+        }
+        showPicker(preference);
+        return true;
+    }
+
+    private void showPicker(Preference preference) {
+        final Context context = preference.getContext();
+        final String setting = setting();
+        final int original = current();
+        final int count = isColor() ? COLOR_NAMES.length : SHAPE_NAMES.length;
+        final int[] chosen = {clamp(original, count)};
+
+        final LinearLayout row = new LinearLayout(context);
+        row.setOrientation(LinearLayout.HORIZONTAL);
+        row.setGravity(Gravity.CENTER);
+        final int pad = dp(context, 16);
+        row.setPadding(pad, pad, pad, pad);
+
+        final Swatch[] swatches = new Swatch[count];
+        for (int i = 0; i < count; i++) {
+            final int index = i;
+            final Swatch swatch = new Swatch(context, isColor() ? colorFor(context, i) : 0,
+                    isColor() ? -1 : i);
+            swatch.setContentDescription(context.getString(
+                    (isColor() ? COLOR_NAMES : SHAPE_NAMES)[i]));
+            final LinearLayout.LayoutParams lp =
+                    new LinearLayout.LayoutParams(dp(context, 44), dp(context, 44));
+            lp.setMarginStart(dp(context, 4));
+            lp.setMarginEnd(dp(context, 4));
+            row.addView(swatch, lp);
+            swatches[i] = swatch;
+            swatch.setOnClickListener(v -> {
+                chosen[0] = index;
+                for (int j = 0; j < count; j++) swatches[j].setSelectedRing(j == index);
+                MotionAssistKeys.put(mContext, setting, index); // live preview
+            });
+            swatch.setSelectedRing(i == chosen[0]);
+        }
+
+        new AlertDialog.Builder(context)
+                .setTitle(preference.getTitle())
+                .setView(row)
+                .setPositiveButton(R.string.motion_assist_save, (d, w) -> {
+                    MotionAssistKeys.put(mContext, setting, chosen[0]);
+                    preference.setSummary(getSummary());
+                })
+                .setNegativeButton(android.R.string.cancel,
+                        (d, w) -> MotionAssistKeys.put(mContext, setting, original))
+                .setOnCancelListener(d -> MotionAssistKeys.put(mContext, setting, original))
+                .show();
+    }
+
+    private boolean isColor() {
+        return KEY_COLOR.equals(getPreferenceKey());
+    }
+
+    private boolean isShape() {
+        return KEY_SHAPE.equals(getPreferenceKey());
+    }
+
+    private String setting() {
+        return isColor() ? MotionAssistKeys.COLOR : MotionAssistKeys.SHAPE;
+    }
+
+    private int current() {
+        return MotionAssistKeys.get(mContext, setting(), 0);
+    }
+
+    private static int clamp(int value, int count) {
+        return value < 0 || value >= count ? 0 : value;
+    }
+
+    private static int dp(Context context, int dp) {
+        return Math.round(TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, dp,
+                context.getResources().getDisplayMetrics()));
+    }
+
+    private static int colorFor(Context context, int index) {
+        final boolean night = (context.getResources().getConfiguration().uiMode
+                & Configuration.UI_MODE_NIGHT_MASK) == Configuration.UI_MODE_NIGHT_YES;
+        if (index == MotionAssistKeys.COLOR_SYSTEM) {
+            return context.getColor(night
+                    ? android.R.color.system_accent1_200 : android.R.color.system_accent1_600);
+        }
+        if (index == MotionAssistKeys.COLOR_CONTRAST) {
+            return night ? Color.rgb(0xF4, 0xF4, 0xF4) : Color.rgb(0x1C, 0x1C, 0x1C);
+        }
+        return PALETTE[index];
+    }
+
+    /** A colour dot or a shape glyph, with a ring when selected. */
+    private static final class Swatch extends View {
+        private final Paint mPaint = new Paint(Paint.ANTI_ALIAS_FLAG);
+        private final Path mPath = new Path();
+        private final int mShape;
+
+        Swatch(Context context, int color, int shape) {
+            super(context);
+            mShape = shape;
+            final TypedValue value = new TypedValue();
+            context.getTheme().resolveAttribute(android.R.attr.textColorPrimary, value, true);
+            mPaint.setColor(shape >= 0 ? context.getColor(value.resourceId) : color);
+            setClickable(true);
+            setFocusable(true);
+        }
+
+        void setSelectedRing(boolean selected) {
+            setSelected(selected);
+            if (!selected) {
+                setBackground(null);
+                return;
+            }
+            final GradientDrawable ring = new GradientDrawable();
+            ring.setShape(GradientDrawable.OVAL);
+            ring.setStroke(dp(getContext(), 2),
+                    getContext().getColor(android.R.color.system_accent1_400));
+            setBackground(ring);
+        }
+
+        @Override
+        protected void onDraw(Canvas canvas) {
+            super.onDraw(canvas);
+            final float cx = getWidth() / 2f;
+            final float cy = getHeight() / 2f;
+            final float r = Math.min(getWidth(), getHeight()) * 0.32f;
+            switch (mShape) {
+                case 1:
+                    canvas.drawRoundRect(cx - r, cy - r, cx + r, cy + r, r * 0.38f, r * 0.38f,
+                            mPaint);
+                    break;
+                case 2:
+                case 3:
+                    mPath.rewind();
+                    final int corners = mShape == 2 ? 5 : 4;
+                    for (int i = 0; i < corners; i++) {
+                        final double a = Math.toRadians(-90.0 + 360.0 / corners * i);
+                        final float px = cx + (float) (r * Math.cos(a));
+                        final float py = cy + (float) (r * Math.sin(a));
+                        if (i == 0) mPath.moveTo(px, py); else mPath.lineTo(px, py);
+                    }
+                    mPath.close();
+                    canvas.drawPath(mPath, mPaint);
+                    break;
+                default:
+                    canvas.drawCircle(cx, cy, r, mPaint);
+            }
+        }
+    }
+}
diff --git a/src/com/android/settings/accessibility/motionassist/MotionAssistCustomizeSettings.java b/src/com/android/settings/accessibility/motionassist/MotionAssistCustomizeSettings.java
new file mode 100644
index 0000000..83b5c7d
--- /dev/null
+++ b/src/com/android/settings/accessibility/motionassist/MotionAssistCustomizeSettings.java
@@ -0,0 +1,38 @@
+/*
+ * SPDX-FileCopyrightText: 2026 petronijevicm
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package com.android.settings.accessibility.motionassist;
+
+import android.app.settings.SettingsEnums;
+
+import com.android.settings.R;
+import com.android.settings.dashboard.DashboardFragment;
+import com.android.settings.search.BaseSearchIndexProvider;
+import com.android.settingslib.search.SearchIndexable;
+
+/** Appearance and behaviour of the cues. Changes apply live while cues are showing. */
+@SearchIndexable(forTarget = SearchIndexable.ALL & ~SearchIndexable.ARC)
+public class MotionAssistCustomizeSettings extends DashboardFragment {
+
+    private static final String TAG = "MotionAssistCustomize";
+
+    @Override
+    public int getMetricsCategory() {
+        return SettingsEnums.ACCESSIBILITY;
+    }
+
+    @Override
+    protected String getLogTag() {
+        return TAG;
+    }
+
+    @Override
+    protected int getPreferenceScreenResId() {
+        return R.xml.motion_assist_customize_settings;
+    }
+
+    public static final BaseSearchIndexProvider SEARCH_INDEX_DATA_PROVIDER =
+            new BaseSearchIndexProvider(R.xml.motion_assist_customize_settings);
+}
diff --git a/src/com/android/settings/accessibility/motionassist/MotionAssistDisplayAreaPreferenceController.java b/src/com/android/settings/accessibility/motionassist/MotionAssistDisplayAreaPreferenceController.java
new file mode 100644
index 0000000..fd6e3f7
--- /dev/null
+++ b/src/com/android/settings/accessibility/motionassist/MotionAssistDisplayAreaPreferenceController.java
@@ -0,0 +1,49 @@
+/*
+ * SPDX-FileCopyrightText: 2026 petronijevicm
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package com.android.settings.accessibility.motionassist;
+
+import android.content.Context;
+
+import androidx.preference.ListPreference;
+import androidx.preference.Preference;
+
+import com.android.settings.core.BasePreferenceController;
+
+/** Where cues appear: side bands (keeps the reading area clear) or the whole screen. */
+public class MotionAssistDisplayAreaPreferenceController extends BasePreferenceController
+        implements Preference.OnPreferenceChangeListener {
+
+    public MotionAssistDisplayAreaPreferenceController(Context context, String key) {
+        super(context, key);
+    }
+
+    @Override
+    public int getAvailabilityStatus() {
+        return AVAILABLE;
+    }
+
+    @Override
+    public void updateState(Preference preference) {
+        super.updateState(preference);
+        if (preference instanceof ListPreference) {
+            final ListPreference list = (ListPreference) preference;
+            list.setValue(String.valueOf(MotionAssistKeys.get(mContext, MotionAssistKeys.MODE, 0)));
+            list.setSummary(list.getEntry());
+        }
+    }
+
+    @Override
+    public boolean onPreferenceChange(Preference preference, Object newValue) {
+        final int mode = Integer.parseInt((String) newValue);
+        if (!MotionAssistKeys.put(mContext, MotionAssistKeys.MODE, mode)) return false;
+        if (preference instanceof ListPreference) {
+            final ListPreference list = (ListPreference) preference;
+            final int index = list.findIndexOfValue((String) newValue);
+            if (index >= 0) list.setSummary(list.getEntries()[index]);
+        }
+        return true;
+    }
+}
diff --git a/src/com/android/settings/accessibility/motionassist/MotionAssistEntryPreferenceController.java b/src/com/android/settings/accessibility/motionassist/MotionAssistEntryPreferenceController.java
new file mode 100644
index 0000000..0725170
--- /dev/null
+++ b/src/com/android/settings/accessibility/motionassist/MotionAssistEntryPreferenceController.java
@@ -0,0 +1,34 @@
+/*
+ * SPDX-FileCopyrightText: 2026 petronijevicm
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package com.android.settings.accessibility.motionassist;
+
+import android.content.Context;
+
+import com.android.settings.R;
+import com.android.settings.core.BasePreferenceController;
+
+/** Entry in the Accessibility list; its summary shows whether cues are on, off or automatic. */
+public class MotionAssistEntryPreferenceController extends BasePreferenceController {
+
+    public MotionAssistEntryPreferenceController(Context context, String key) {
+        super(context, key);
+    }
+
+    @Override
+    public int getAvailabilityStatus() {
+        return AVAILABLE;
+    }
+
+    @Override
+    public CharSequence getSummary() {
+        if (!MotionAssistKeys.isOn(mContext, MotionAssistKeys.ENABLED, false)) {
+            return mContext.getString(R.string.motion_assist_state_off);
+        }
+        return mContext.getString(MotionAssistKeys.isOn(mContext, MotionAssistKeys.VEHICLE_AUTO, false)
+                ? R.string.motion_assist_state_auto
+                : R.string.motion_assist_state_on);
+    }
+}
diff --git a/src/com/android/settings/accessibility/motionassist/MotionAssistKeys.java b/src/com/android/settings/accessibility/motionassist/MotionAssistKeys.java
new file mode 100644
index 0000000..29c34b3
--- /dev/null
+++ b/src/com/android/settings/accessibility/motionassist/MotionAssistKeys.java
@@ -0,0 +1,47 @@
+/*
+ * SPDX-FileCopyrightText: 2026 petronijevicm
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package com.android.settings.accessibility.motionassist;
+
+import android.content.Context;
+import android.provider.Settings;
+
+/**
+ * Settings.Secure keys for Motion Assist, shared with SystemUI's MotionAssistController.
+ * Keep names and defaults in sync with that class.
+ */
+final class MotionAssistKeys {
+
+    static final String ENABLED = "motion_assist_enabled";
+    static final String VEHICLE_AUTO = "motion_assist_vehicle_auto";
+    static final String COLOR = "motion_assist_color";
+    static final String SHAPE = "motion_assist_shape";
+    static final String OPACITY = "motion_assist_opacity";
+    static final String SIZE = "motion_assist_size";
+    static final String MOVEMENT = "motion_assist_movement";
+    static final String RESPONSIVENESS = "motion_assist_responsiveness";
+    static final String MODE = "motion_assist_mode";
+    static final String SHUFFLE = "motion_assist_randomize";
+    static final String SMOOTH = "motion_assist_smooth_animation";
+    /** Watched by SystemUI's Quick Settings auto-add list. */
+    static final String ADD_TILE = "motion_assist_add_qs_tile";
+
+    static final int COLOR_SYSTEM = 0;
+    static final int COLOR_CONTRAST = 5;
+
+    private MotionAssistKeys() {}
+
+    static int get(Context context, String key, int def) {
+        return Settings.Secure.getInt(context.getContentResolver(), key, def);
+    }
+
+    static boolean put(Context context, String key, int value) {
+        return Settings.Secure.putInt(context.getContentResolver(), key, value);
+    }
+
+    static boolean isOn(Context context, String key, boolean def) {
+        return get(context, key, def ? 1 : 0) == 1;
+    }
+}
diff --git a/src/com/android/settings/accessibility/motionassist/MotionAssistSettings.java b/src/com/android/settings/accessibility/motionassist/MotionAssistSettings.java
new file mode 100644
index 0000000..db7756b
--- /dev/null
+++ b/src/com/android/settings/accessibility/motionassist/MotionAssistSettings.java
@@ -0,0 +1,38 @@
+/*
+ * SPDX-FileCopyrightText: 2026 petronijevicm
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package com.android.settings.accessibility.motionassist;
+
+import android.app.settings.SettingsEnums;
+
+import com.android.settings.R;
+import com.android.settings.dashboard.DashboardFragment;
+import com.android.settings.search.BaseSearchIndexProvider;
+import com.android.settingslib.search.SearchIndexable;
+
+/** Main Motion Assist screen: tutorial, on/off, auto start, Quick Settings tile. */
+@SearchIndexable(forTarget = SearchIndexable.ALL & ~SearchIndexable.ARC)
+public class MotionAssistSettings extends DashboardFragment {
+
+    private static final String TAG = "MotionAssistSettings";
+
+    @Override
+    public int getMetricsCategory() {
+        return SettingsEnums.ACCESSIBILITY;
+    }
+
+    @Override
+    protected String getLogTag() {
+        return TAG;
+    }
+
+    @Override
+    protected int getPreferenceScreenResId() {
+        return R.xml.motion_assist_settings;
+    }
+
+    public static final BaseSearchIndexProvider SEARCH_INDEX_DATA_PROVIDER =
+            new BaseSearchIndexProvider(R.xml.motion_assist_settings);
+}
diff --git a/src/com/android/settings/accessibility/motionassist/MotionAssistSliderPreferenceController.java b/src/com/android/settings/accessibility/motionassist/MotionAssistSliderPreferenceController.java
new file mode 100644
index 0000000..226af4e
--- /dev/null
+++ b/src/com/android/settings/accessibility/motionassist/MotionAssistSliderPreferenceController.java
@@ -0,0 +1,85 @@
+/*
+ * SPDX-FileCopyrightText: 2026 petronijevicm
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package com.android.settings.accessibility.motionassist;
+
+import android.content.Context;
+
+import androidx.preference.Preference;
+import androidx.preference.PreferenceScreen;
+
+import com.android.settings.core.SliderPreferenceController;
+import com.android.settingslib.widget.SliderPreference;
+
+/**
+ * Sliders backed by a Motion Assist Settings.Secure key: opacity, size, movement and
+ * responsiveness. The preference key selects the setting and its range.
+ */
+public class MotionAssistSliderPreferenceController extends SliderPreferenceController {
+
+    private final String mSetting;
+    private final int mMin;
+    private final int mMax;
+    private final int mDefault;
+
+    public MotionAssistSliderPreferenceController(Context context, String key) {
+        super(context, key);
+        switch (key) {
+            case "motion_assist_opacity":
+                mSetting = MotionAssistKeys.OPACITY; mMin = 10; mMax = 100; mDefault = 60;
+                break;
+            case "motion_assist_size":
+                mSetting = MotionAssistKeys.SIZE; mMin = 50; mMax = 200; mDefault = 100;
+                break;
+            case "motion_assist_movement":
+                mSetting = MotionAssistKeys.MOVEMENT; mMin = 25; mMax = 200; mDefault = 100;
+                break;
+            case "motion_assist_responsiveness":
+                mSetting = MotionAssistKeys.RESPONSIVENESS; mMin = 0; mMax = 100; mDefault = 70;
+                break;
+            default:
+                mSetting = null; mMin = 0; mMax = 100; mDefault = 0;
+        }
+    }
+
+    @Override
+    public int getAvailabilityStatus() {
+        return mSetting != null ? AVAILABLE : UNSUPPORTED_ON_DEVICE;
+    }
+
+    @Override
+    public void displayPreference(PreferenceScreen screen) {
+        super.displayPreference(screen);
+        final Preference preference = screen.findPreference(getPreferenceKey());
+        if (preference instanceof SliderPreference) {
+            final SliderPreference slider = (SliderPreference) preference;
+            slider.setMin(mMin);
+            slider.setMax(mMax);
+            slider.setUpdatesContinuously(true);
+        }
+    }
+
+    @Override
+    public int getSliderPosition() {
+        if (mSetting == null) return mMin;
+        return Math.max(mMin, Math.min(mMax, MotionAssistKeys.get(mContext, mSetting, mDefault)));
+    }
+
+    @Override
+    public boolean setSliderPosition(int position) {
+        return mSetting != null && MotionAssistKeys.put(mContext, mSetting,
+                Math.max(mMin, Math.min(mMax, position)));
+    }
+
+    @Override
+    public int getMax() {
+        return mMax;
+    }
+
+    @Override
+    public int getMin() {
+        return mMin;
+    }
+}
diff --git a/src/com/android/settings/accessibility/motionassist/MotionAssistTilePreferenceController.java b/src/com/android/settings/accessibility/motionassist/MotionAssistTilePreferenceController.java
new file mode 100644
index 0000000..ace8ee8
--- /dev/null
+++ b/src/com/android/settings/accessibility/motionassist/MotionAssistTilePreferenceController.java
@@ -0,0 +1,83 @@
+/*
+ * SPDX-FileCopyrightText: 2026 petronijevicm
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package com.android.settings.accessibility.motionassist;
+
+import android.content.Context;
+import android.graphics.drawable.Drawable;
+import android.provider.Settings;
+import android.text.SpannableStringBuilder;
+import android.text.Spanned;
+import android.text.TextUtils;
+import android.text.style.ImageSpan;
+
+import androidx.appcompat.app.AlertDialog;
+import androidx.preference.Preference;
+
+import com.android.settings.R;
+import com.android.settings.core.BasePreferenceController;
+
+/** "Add tile to Quick Settings": adds the tile, then explains where to find it. */
+public class MotionAssistTilePreferenceController extends BasePreferenceController {
+
+    private static final String QS_TILES = "sysui_qs_tiles";
+    private static final String TILE_SPEC = "motion_assist";
+    private static final String ICON_PLACEHOLDER = "%s";
+
+    public MotionAssistTilePreferenceController(Context context, String key) {
+        super(context, key);
+    }
+
+    @Override
+    public int getAvailabilityStatus() {
+        return AVAILABLE;
+    }
+
+    @Override
+    public boolean handlePreferenceTreeClick(Preference preference) {
+        if (!TextUtils.equals(preference.getKey(), getPreferenceKey())) {
+            return super.handlePreferenceTreeClick(preference);
+        }
+        addTile();
+        showHowToUse(preference.getContext());
+        return true;
+    }
+
+    private void addTile() {
+        // SystemUI's auto-add list reacts to this key; writing the tile list covers the case
+        // where the tile was added and removed before.
+        MotionAssistKeys.put(mContext, MotionAssistKeys.ADD_TILE, 1);
+        final String tiles = Settings.Secure.getString(mContext.getContentResolver(), QS_TILES);
+        if (TextUtils.isEmpty(tiles)) {
+            return; // SystemUI still uses its defaults, which include the tile
+        }
+        for (String spec : tiles.split(",")) {
+            if (TILE_SPEC.equals(spec.trim())) return;
+        }
+        Settings.Secure.putString(mContext.getContentResolver(), QS_TILES, tiles + "," + TILE_SPEC);
+    }
+
+    private void showHowToUse(Context context) {
+        final String template = context.getString(R.string.motion_assist_tile_dialog_edit);
+        final SpannableStringBuilder text = new SpannableStringBuilder(
+                context.getString(R.string.motion_assist_tile_dialog_use));
+        text.append("\n\n");
+        final int start = text.length() + template.indexOf(ICON_PLACEHOLDER);
+        text.append(template);
+        final Drawable pencil = context.getDrawable(R.drawable.ic_motion_assist_edit);
+        if (pencil != null && template.contains(ICON_PLACEHOLDER)) {
+            final int size = Math.round(context.getResources().getDisplayMetrics().density * 18);
+            pencil.setBounds(0, 0, size, size);
+            text.setSpan(new ImageSpan(pencil, ImageSpan.ALIGN_BOTTOM), start,
+                    start + ICON_PLACEHOLDER.length(), Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);
+        }
+        new AlertDialog.Builder(context)
+                .setIcon(R.drawable.ic_motion_assist)
+                .setTitle(R.string.motion_assist_tile_dialog_title)
+                .setMessage(text)
+                .setPositiveButton(R.string.motion_assist_got_it, null)
+                .show();
+    }
+}
diff --git a/src/com/android/settings/accessibility/motionassist/MotionAssistTogglePreferenceController.java b/src/com/android/settings/accessibility/motionassist/MotionAssistTogglePreferenceController.java
new file mode 100644
index 0000000..c7425c2
--- /dev/null
+++ b/src/com/android/settings/accessibility/motionassist/MotionAssistTogglePreferenceController.java
@@ -0,0 +1,63 @@
+/*
+ * SPDX-FileCopyrightText: 2026 petronijevicm
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package com.android.settings.accessibility.motionassist;
+
+import android.content.Context;
+
+import com.android.settings.R;
+import com.android.settings.core.TogglePreferenceController;
+
+/**
+ * On/off switches backed by a Motion Assist Settings.Secure key. The preference key names the
+ * setting, so one controller serves the main switch, auto start, shuffle and smooth animation.
+ */
+public class MotionAssistTogglePreferenceController extends TogglePreferenceController {
+
+    public MotionAssistTogglePreferenceController(Context context, String key) {
+        super(context, key);
+    }
+
+    @Override
+    public int getAvailabilityStatus() {
+        return settingKey() != null ? AVAILABLE : UNSUPPORTED_ON_DEVICE;
+    }
+
+    @Override
+    public boolean isChecked() {
+        final String setting = settingKey();
+        return setting != null && MotionAssistKeys.isOn(mContext, setting, defaultOn(setting));
+    }
+
+    @Override
+    public boolean setChecked(boolean isChecked) {
+        final String setting = settingKey();
+        return setting != null && MotionAssistKeys.put(mContext, setting, isChecked ? 1 : 0);
+    }
+
+    @Override
+    public int getSliceHighlightMenuRes() {
+        return R.string.menu_key_accessibility;
+    }
+
+    private String settingKey() {
+        switch (getPreferenceKey()) {
+            case "motion_assist_main_switch":
+                return MotionAssistKeys.ENABLED;
+            case "motion_assist_auto_start":
+                return MotionAssistKeys.VEHICLE_AUTO;
+            case "motion_assist_shuffle":
+                return MotionAssistKeys.SHUFFLE;
+            case "motion_assist_smooth":
+                return MotionAssistKeys.SMOOTH;
+            default:
+                return null;
+        }
+    }
+
+    private static boolean defaultOn(String setting) {
+        return MotionAssistKeys.SMOOTH.equals(setting);
+    }
+}
-- 
2.43.0
